Download patch files from mirrors

Askaholic opened the pull request:
Closes #2433

I also made it so that before trying to download ForgedAlliance.exe, the client will try to use the currently installed version and just patch the version number as the majority of updates don't actually change the exe other than the version number. If the checksum of the patched file doesn't match then it will download the exe as usual.

![image](https://user-images.githubusercontent.com/17941539/144781983-27fac522-577b-43c7-90a0-6d2f5d393a1b.png)

Askaholic committed:
Download game files using a list of mirrors

Askaholic committed:
Get mirror urls from preferences

Askaholic committed:
Try to patch ForgedAlliance.exe version number before downloading

Sheikah45 commented on src/main/java/com/faforever/client/io/ChecksumMismatchException.java:
```diff
@@ -0,0 +1,22 @@
+
+import lombok.Getter;
+
+public class ChecksumMismatchException extends Exception {
+  @Getter
+  private final Object item;
+  @Getter
+  private final String expected;
+  @Getter
+  private final String actual;
+
+  public ChecksumMismatchException(Object item, String expected, String actual) {
+    this.item = item;
+    this.expected = expected;
+    this.actual = actual;
+  }
```
```suggestion
@Getter
@RequiredArgsConstructor
public class ChecksumMismatchException extends Exception {
  private final Object item;
  private final String expected;
  private final String actual;
```

Sheikah45 replied:
Is object really the most restrictive type? Couldn't it be something like File or Path?

Sheikah45 commented on src/main/java/com/faforever/client/io/DownloadService.java:
```diff
@@ -1,41 +1,91 @@
+      } catch (FileNotFoundException e) {
+        log.warn("Could not download file at " + mirrorUrl);
+      } catch (ChecksumMismatchException e) {
+        log.warn(String.format("Checksum downloaded from %s did not match. Expected %s got %s", mirrorUrl, e.getExpected(), e.getActual()));
```
You don't need String.format here. Also if you have the toString method the same as this method then use that
```suggestion
        log.warn("Checksum did not match from {}", mirrorUrl, e);
```

Sheikah45 commented on src/main/java/com/faforever/client/io/DownloadService.java:
```diff
@@ -1,41 +1,91 @@
+      } catch (ChecksumMismatchException e) {
+        log.warn(String.format("Checksum downloaded from %s did not match. Expected %s got %s", mirrorUrl, e.getExpected(), e.getActual()));
+      } catch (IOException e) {
+        log.warn("Could not download file at " + mirrorUrl, e);
```
```suggestion
        log.warn("Could not download file at {}" , mirrorUrl, e);
```

Sheikah45 commented on src/main/java/com/faforever/client/io/DownloadService.java:
```diff
@@ -1,41 +1,91 @@
+   */
+  public void downloadFile(URL url, Path targetFile, ByteCountListener progressListener, String md5sum) throws IOException, NoSuchAlgorithmException, ChecksumMismatchException {
     Path tempFile = Files.createTempFile(targetFile.getParent(), "download", null);
+    log.debug("Downloading file " + url + " to " + tempFile);
```
```suggestion
    log.debug("Downloading {} to {}", url, tempFile);
```

Sheikah45 replied:
This still needs to be changed

Sheikah45 commented on src/main/java/com/faforever/client/io/DownloadService.java:
```diff
@@ -46,4 +96,25 @@ public void downloadFile(URL url, Path targetFile, ByteCountListener progressLis
+  public List<URL> getMirrorsFor(URL url) {
+    return preferencesService.getPreferences().getMirror().getMirrorURLs().stream()
+        .map(mirror -> getMirrorURL(mirror, url))
+        .filter(Optional::isPresent)
+        .map(Optional::get)
```
```suggestion
        .flatMap(Optional::stream)
```

Sheikah45 commented on src/main/java/com/faforever/client/patch/GameUpdaterImpl.java:
```diff
@@ -102,16 +102,20 @@ public GameUpdater addFeaturedModUpdater(FeaturedModUpdater featuredModUpdater)
+            } else if (throwable.getCause() instanceof ChecksumMismatchException) {
+              throw new CompletionException(throwable.getCause());
+            } else {
+              log.info("Launching anyway", throwable);
```
```suggestion
              log.info("Ignored error while updating featured mod", throwable);
```

Sheikah45 commented on src/main/java/com/faforever/client/preferences/MirrorPrefs.java:
```diff
@@ -0,0 +1,20 @@
+import java.net.URI;
+
+public class MirrorPrefs {
+  private final ListProperty<URI> mirrorURLs;
+
+  public MirrorPrefs() {
+    mirrorURLs = new SimpleListProperty<>(FXCollections.observableArrayList());
+  }
+
+  public ObservableList<URI> getMirrorURLs() {
+    return mirrorURLs.get();
```
```suggestion
  private final ObservableList<URI> mirrorURLs;

  public MirrorPrefs() {
    mirrorURLs = FXCollections.observableArrayList();
  }

  public ObservableList<URI> getMirrorURLs() {
    return mirrorURLs;
```

Sheikah45 commented on src/main/java/com/faforever/client/preferences/ui/SettingsController.java:
```diff
@@ -310,11 +313,19 @@ public void initialize() {
+    mirrorURLsListView.setItems(preferencesService.getPreferences().getMirror().getMirrorURLs());
+    mirrorURLsListView.setCellFactory(param -> uiService.<RemovableListCellController<URI>>loadFxml("theme/settings/removable_cell.fxml"));
+    mirrorURLsListView.getItems().addListener((ListChangeListener<URI>) c -> preferencesService.storeInBackground());
+    mirrorURLsListView.managedProperty().bind(mirrorURLsListView.visibleProperty());
```
```suggestion
   JavaFxUtil.bindManagedToVisible(mirrorURLsListView);
```

Sheikah45 commented on src/main/java/com/faforever/client/preferences/ui/SettingsController.java:
```diff
@@ -310,11 +313,19 @@ public void initialize() {
+    mirrorURLsListView.setCellFactory(param -> uiService.<RemovableListCellController<URI>>loadFxml("theme/settings/removable_cell.fxml"));
+    mirrorURLsListView.getItems().addListener((ListChangeListener<URI>) c -> preferencesService.storeInBackground());
+    mirrorURLsListView.managedProperty().bind(mirrorURLsListView.visibleProperty());
+    mirrorURLsListView.visibleProperty().bind(Bindings.createBooleanBinding(() -> !mirrorURLsListView.getItems().isEmpty(), mirrorURLsListView.getItems()));
```
If this is ever updated not through the UI then this needs to be a listener and not a binding

Askaholic replied:
Yea idk, I just copy/pasted from the autojoinchannel listview

Sheikah45 commented on src/main/java/com/faforever/client/preferences/ui/SettingsController.java:
```diff
@@ -589,5 +600,21 @@ public void onPatchGamePrefsForMultipleInstances() {
+
+  public void onAddMirrorURL() {
+    URI uri = URI.create(mirrorURITextField.getText());
+    if (mirrorURITextField.getText().isEmpty() || mirrorURLsListView.getItems().contains(uri)) {
```
```suggestion
    if (mirrorURITextField.getText().isBlank() || mirrorURLsListView.getItems().contains(uri)) {
```

Sheikah45 commented on src/main/java/com/faforever/client/preferences/ui/SettingsController.java:
```diff
@@ -589,5 +600,21 @@ public void onPatchGamePrefsForMultipleInstances() {
+    }
+    try {
+      uri.toURL();
+    } catch (Exception e) {
```
```suggestion
    } catch (MalformedURLException e) {
```

Askaholic replied:
No, I had that and then it threw an IllegalArgumentException

Sheikah45 replied:
Then this should onlt check for MalformedURLException and IllegalArgumentException

Also the try should really be around the whole method body since URI.create can also throw the IllegalArgumentException if it doesn't match the spec. If you want to check both the URL and URI at the same time I would just do
`URI uri = new URL(text).toURI()`
and add the try catch around that

Sheikah45 reviewed (changes requested):
Do not make changes to the messages_XX.properties files those are autoupdatedby a third party. So all those changes should be reverted

Sheikah45 commented on src/test/java/com/faforever/client/patch/DownloadServiceTest.java:
```diff
@@ -0,0 +1,78 @@
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.mockito.Mockito.when;
+
+public class DownloadServiceTest extends ServiceTest {
```
Need a test that the downloadFileWithMirrors rotates through the mirrors properly as well

Askaholic committed:
Refactor from review suggestions

Askaholic committed:
Move DownloadServiceTest to correct package

Askaholic committed:
Add tests for downloadFileWithMirrors

Sheikah45 commented on src/main/java/com/faforever/client/io/ChecksumMismatchException.java:
```diff
@@ -0,0 +1,17 @@
+@Getter
+@RequiredArgsConstructor
+public class ChecksumMismatchException extends Exception {
+  private final Object item;
```
Can this be more specific than object?

Sheikah45 commented on src/main/java/com/faforever/client/io/ChecksumMismatchException.java:
```diff
@@ -0,0 +1,17 @@
+public class ChecksumMismatchException extends Exception {
+  private final Object item;
+  private final String expected;
+  private final String actual;
```
Also sorry just realized they are all private final 
```suggestion
@Value
public class ChecksumMismatchException extends Exception {
  Object item;
  String expected;
  String actual;
```

Sheikah45 commented on src/main/resources/i18n/messages_cs.properties:
```diff
@@ -758,4 +758,7 @@ chat.userContext.report = Nahlásit hráče
-game.generateMap.more = Více
\ No newline at end of file
+game.generateMap.more = Více
+settings.data.mirrorURLs=Download Mirrors
+settings.data.mirrorURLs.description=Add any mirrors here that you wish to use for downloading game patches instead of downloading from the main FAF server. This can help prevent extremely long download times.
+settings.data.mirrorURLPrompt=https://faf-mirror.example.com/
```
These need to be removed from all but the message.properties files

Askaholic replied:
Weird I thought I did that already… will take another look.

Sheikah45 replied:
Ah you probably did I wiped some changes with the rebase but thought I fixed them all. So I will go ahead and fix this

Sheikah45 replied:
Since I caused it

Sheikah45 reviewed (changes requested):
Need to revert the changes to the other messages_x.properties files

Sheikah45 committed:
Remove changes to messages files

Sheikah45 commented on src/main/java/com/faforever/client/patch/GameUpdaterImpl.java:
```diff
@@ -82,18 +83,22 @@ public GameUpdater addFeaturedModUpdater(FeaturedModUpdater featuredModUpdater)
+            if (throwable.getCause() instanceof AccessDeniedException) {
+              throw new UnsupportedOperationException("Unable to patch Forged Alliance to the required version " +
+                  "due to conflicting version running", throwable);
+            } else if (throwable.getCause() instanceof ChecksumMismatchException) {
+              throw new CompletionException(throwable.getCause());
```
Ah just realized this will only rethrow when the player has allowReplaysWhileinGame set which is probably not intended.

I can fix it though as there are a few things to clean up I have been meaning to do

Sheikah45 committed:
Ensure ChecksumMismatchException or UnsupportedOperationExceptions throw

Askaholic committed:
Replace binding with listener

Askaholic committed:
Clean up log string formatting

Sheikah45 commented on src/main/java/com/faforever/client/io/DownloadService.java:
```diff
@@ -1,49 +1,119 @@
+        downloadFile(mirrorUrl, targetFile, progressListener, md5sum);
+        return;
+      } catch (FileNotFoundException e) {
+        log.warn("Could not download file at {}", mirrorUrl);
```
log the exception here please. Also this can be combined with the IOexception catch no need to have 2 that are the same

Askaholic replied:
Oh yea the weird thing here is that URL will throw this `FileNotFoundException` if it encounters an http error when trying to open the URL. That's why I separated it out, because it's specifically catching the http errors. I should probably make it log on INFO or something though because I would expect this to happen all the time given that the mirrors may be out of date and not contain all needed files.

Sheikah45 replied:
Maybe but if the error message and log levels are the same then there is no need to separate them out since they are performing the same action

Askaholic replied:
I changed it to info

Sheikah45 commented on src/main/java/com/faforever/client/io/DownloadService.java:
```diff
@@ -1,49 +1,119 @@
+      // NOTE: It is crucial that we verify the checksum before using the file when downloading from mirrors! We don't
+      // want to be running unverified executables!
+      String checksum = DatatypeConverter.printHexBinary(messageDigest.digest()).toLowerCase();
+      if (!Objects.equals(md5sum, checksum)) {
+        throw new ChecksumMismatchException(url, checksum, md5sum);
+      }
```
Can we add a test that exercises this path?

Sheikah45 commented on src/main/java/com/faforever/client/io/DownloadService.java:
```diff
@@ -1,49 +1,119 @@
+  /*
+   * Get URLs to the same file on all available mirrors.
+   */
+  public List<URL> getMirrorsFor(URL url) {
+    return preferencesService.getPreferences().getMirror().getMirrorURLs().stream()
+        .map(mirror -> getMirrorURL(mirror, url))
+        .flatMap(Optional::stream)
+        .collect(Collectors.toList());
+  }
+
+  public Optional<URL> getMirrorURL(URI mirror, URL url) {
+      URI uri = mirror.resolve(url.getPath());
+      try {
+        return Optional.of(uri.toURL());
+      } catch (MalformedURLException e) {
+        log.warn("Failed to create URL from URI: {}", uri);
+        return Optional.empty();
+      }
```
Can we add tests for these as well?

Askaholic replied:
I already added tests for these

Askaholic replied:
Not sure why it says that line is uncovered because I definitely already have a test that should hit the 'empty' path here. Maybe my test class isn't being picked up for some reason?

Sheikah45 replied:
Could be I will double check and look at the tests. Codecov can be spotty at times

Askaholic replied:
I do get this error when trying to run the test through IDEA:
```
> No tests found for given includes: [com.faforever.client.io.DownloadServiceTest.testDownloadFile](filter.includeTestsMatching)
```

Sheikah45 commented on src/main/java/com/faforever/client/preferences/ui/SettingsController.java:
```diff
@@ -589,5 +602,26 @@ public void onPatchGamePrefsForMultipleInstances() {
     }
   }
+
+  public void onAddMirrorURL() {
+    String text = mirrorURITextField.getText();
+    if (text.isBlank()) {
+      return;
+    }
+
+    URI uri = null;
+    try {
+      uri = new URL(text).toURI();
+    } catch (URISyntaxException | MalformedURLException e) {
+      log.debug("Failed to add invalid URL: {}", text, e);
+    }
+
+    if (mirrorURLsListView.getItems().contains(uri)) {
+      return;
+    }
+    preferencesService.getPreferences().getMirror().getMirrorURLs().add(uri);
+    preferencesService.storeInBackground();
+    mirrorURITextField.clear();
+  }
```
Test please

Sheikah45 reviewed (changes requested):
A few tests to add then we should be good

Askaholic committed:
Add more tests

Sheikah45 commented on src/test/java/com/faforever/client/io/DownloadServiceTest.java:
```diff
@@ -129,4 +133,35 @@ public void testDownloadWithMirrors() throws Exception {
+    assertEquals(data, DATA);
+  }
+
+  @Test(expected = ChecksumMismatchException.class)
+  public void testDownloadFileBadChecksum() throws Exception {
+    URL url = mock(URL.class);
+    URLConnection urlConnection = mock(URLConnection.class);
+
+    byte[] DATA = "some content".getBytes();
+    when(url.openStream()).thenReturn(new ByteArrayInputStream(DATA));
+    when(urlConnection.getContentLength()).thenReturn(DATA.length);
+    when(url.openConnection()).thenReturn(urlConnection);
+
+    Path temp = Files.createTempFile("download", ".dat");
+    instance.downloadFile(url, temp, (a, b) -> {}, "00000000000000000000000000000000");
+  }
```
It is better to wrap the instance.downloadFile inside of an Assertions.assertThrows() so that we are checking the specific line we expect to throw the exception rather than the whole method

Askaholic replied:
Yea I wasn’t sure how to do it in Java. I’ll update it

Sheikah45 commented on src/test/java/com/faforever/client/io/DownloadServiceTest.java:
```diff
@@ -0,0 +1,167 @@
+import com.faforever.commons.io.ByteCountListener;
+import javafx.collections.FXCollections;
+import org.assertj.core.api.Assertions;
+import org.junit.Test;
```
This is the wrong test annotation

Sheikah45 commented on src/test/java/com/faforever/client/io/DownloadServiceTest.java:
```diff
@@ -0,0 +1,167 @@
+
+public class DownloadServiceTest extends ServiceTest {
+
+  @Mock(answer = Answers.RETURNS_DEEP_STUBS)
+  private PreferencesService preferencesService;
```
There is no need to use a deep stub here. We have a preferencesBuilder for testing so we can pass complete data

Sheikah45 committed:
Use correct Test annotation and use PreferencesBuilder

Sheikah45 reviewed (commented):
I fixed the tests so that they run now but it looks like some of them are failing so that should be looked into

Askaholic committed:
Get tests to pass

Sheikah45 commented on src/main/java/com/faforever/client/preferences/ui/SettingsController.java:
```diff
@@ -310,11 +316,18 @@ public void initialize() {
+    mirrorURLsListView.setFocusTraversable(false);
+    mirrorURLsListView.setItems(preferencesService.getPreferences().getMirror().getMirrorURLs());
+    mirrorURLsListView.setCellFactory(param -> uiService.<RemovableListCellController<URI>>loadFxml("theme/settings/removable_cell.fxml"));
+    mirrorURLsListView.getItems().addListener((ListChangeListener<URI>) c -> preferencesService.storeInBackground());
+    mirrorURLsListView.getItems().addListener((ListChangeListener<URI>) c -> mirrorURLsListView.setVisible(!mirrorURLsListView.getItems().isEmpty()));
```
rather than two listeners it is probably better to just make both into a function and then call that function in a single listener.

Also will want to use JavaFxUtil.addListener for thread safety just in case.

Askaholic replied:
I switched it to use JavaFxUtil, but I didn't see anywhere in the code that used a method rather than a lambda. Lambda is fine IMO. It even makes it easier to read because the code for the listener is right there where the listener is added.

Sheikah45 commented on src/main/java/com/faforever/client/preferences/ui/SettingsController.java:
```diff
@@ -589,5 +602,30 @@ public void onPatchGamePrefsForMultipleInstances() {
+    try {
+      uri = new URL(text).toURI();
+    } catch (URISyntaxException | MalformedURLException e) {
+      log.debug("Failed to add invalid URL: {}", text, e);
+      return;
```
Probably want a user notification here

Askaholic replied:
It would be nice if there were a way to flash an error message but I think a notification would be a bit overkill. You can tell that there is an error by the fact that the text field is not cleared but it would also be nice to show a small tooltip like message. Not sure if there’s a way to do that in JavaFX.

Sheikah45 replied:
I would just use the addimmeadiateerrornotification

Sheikah45 replied:
Err actually probably just addimmeadiatewarnnotification

Sheikah45 commented on src/main/java/com/faforever/client/preferences/ui/SettingsController.java:
```diff
@@ -589,5 +602,30 @@ public void onPatchGamePrefsForMultipleInstances() {
+      text = text + "/";
+    }
+
+    URI uri = null;
+    try {
+      uri = new URL(text).toURI();
+    } catch (URISyntaxException | MalformedURLException e) {
+      log.debug("Failed to add invalid URL: {}", text, e);
+      return;
+    }
+
+    if (mirrorURLsListView.getItems().contains(uri)) {
+      return;
+    }
+    preferencesService.getPreferences().getMirror().getMirrorURLs().add(uri);
+    preferencesService.storeInBackground();
+    mirrorURITextField.clear();
```
Since you have a return in the catch anyway everything after the catch can just be moved into the try statement, this removes the possibility of uri being null as well since you can move the declaration there as well.

Sheikah45 commented on src/test/java/com/faforever/client/io/DownloadServiceTest.java:
```diff
@@ -0,0 +1,171 @@
+  @Test
+  public void testDownloadFile() throws Exception {
+    Path temp = Files.createTempFile("download", ".dat");
+    instance.downloadFile(SAMPLE_FILE.getURL(), temp, (processed, total) -> {}, "b53227da4280f0e18270f21dd77c91d0");
```
Can we make the correct checksum into a constant as well

Sheikah45 reviewed (changes requested):
Just a few things I missed before

Askaholic committed:
Use JavaFxUtil.addListener

Askaholic committed:
Extract checksum to constant

Askaholic committed:
Add immediate warning notification when URL is invalid

The reasons aren't really going to be localized because they come from 
the execption message.
